diff --git a/mozharness/mozilla/buildbot.py b/mozharness/mozilla/buildbot.py
index e141b09..034ea87 100755
--- a/mozharness/mozilla/buildbot.py
+++ b/mozharness/mozilla/buildbot.py
@@ -16,7 +16,6 @@ sys.path.insert(1, os.path.dirname(sys.path[0]))
 
 from mozharness.base.config import parse_config_file
 from mozharness.base.log import INFO, WARNING, ERROR
-from mozharness.base.log import OutputParser
 
 # BuildbotMixin {{{1
 
diff --git a/mozharness/mozilla/testing/testbase.py b/mozharness/mozilla/testing/testbase.py
index 4f5f727..2f6dc86 100755
--- a/mozharness/mozilla/testing/testbase.py
+++ b/mozharness/mozilla/testing/testbase.py
@@ -192,13 +192,9 @@ Did you run with --create-virtualenv? Is mozinstall in virtualenv_modules?""")
         dirs = self.query_abs_dirs()
         target_dir = dirs.get('abs_app_install_dir',
                               os.path.join(dirs['abs_work_dir'],
-                             'application'))
+                              'application'))
         self.mkdir_p(target_dir)
-        cmd.extend(['--destination', target_dir, self.installer_path])
-        # TODO find out if --source is deprectiated
-        if '--source' in output:
-            for index, value in enumerate(cmd):
-                if value == self.installer_path:
-                    cmd.insert(index, '--source')
+        cmd.extend([self.installer_path,
+                    '--destination', target_dir])
         # TODO we'll need some error checking here
         self.binary_path = self.get_output_from_command(cmd)
diff --git a/scripts/desktop_unittest.py b/scripts/desktop_unittest.py
index 3932dc5..148e507 100755
--- a/scripts/desktop_unittest.py
+++ b/scripts/desktop_unittest.py
@@ -16,14 +16,15 @@ import shutil
 # load modules from parent dir
 sys.path.insert(1, os.path.dirname(sys.path[0]))
 
-from mozharness.base.errors import PythonErrorList, BaseErrorList
-from mozharness.mozilla.testing.errors import TinderBoxPrintRe, TestPassed
+from mozharness.base.errors import BaseErrorList
+from mozharness.mozilla.testing.errors import TinderBoxPrintRe
 from mozharness.base.vcs.vcsbase import MercurialScript
 from mozharness.mozilla.testing.testbase import TestingMixin, testing_config_options
 from mozharness.base.log import OutputParser, WARNING, INFO
 from mozharness.mozilla.buildbot import TBPL_WARNING, TBPL_FAILURE
 from mozharness.mozilla.buildbot import TBPL_SUCCESS, TBPL_STATUS_DICT
 
+
 class DesktopUnittestOutputParser(OutputParser):
     """
     A class that extends OutputParser such that it can parse the number of
@@ -245,7 +246,7 @@ then do not specify to run only specific suites like '--mochitest-suite browser-
                 self.fatal("self.installer_url was found but symbols_url could \
                         not be determined")
         else:
-            self.fatal("self.installer_url was not found in self.config") 
+            self.fatal("self.installer_url was not found in self.config")
         self.info("setting symbols_url as %s" % (symbols_url))
         self.symbols_url = symbols_url
         return self.symbols_url
@@ -270,7 +271,7 @@ then do not specify to run only specific suites like '--mochitest-suite browser-
             else:
                 self.warning("""Suite options for %s could not be determined.
 If you meant to have options for this suite, please make sure they are specified
-in your config under %s_options""" % suite_category, suite_category)
+in your config under %s_options""" % (suite_category, suite_category))
         else:
             self.fatal("""'binary_path' could not be determined.
             This should be something like '/root/path/with/build/application/firefox/firefox-bin'
