diff --git a/configs/talos/windows_config.py b/configs/talos/windows_config.py
index edb9368..cca1ecb 100644
--- a/configs/talos/windows_config.py
+++ b/configs/talos/windows_config.py
@@ -46,4 +46,6 @@ config = {
     "populate_webroot": True,
     # Srsly gly? Ys
     "webroot_extract_cmd": r'''c:/mozilla-build/msys/bin/bash -c "PATH=/c/mozilla-build/msys/bin:$PATH tar zx --strip-components=1 -f '%(tarball)s' --wildcards '**/talos/'"''',
+    "metro_harness_dir": "mozbase/mozrunner/mozrunner/resources",
+    "metro_test_harness_exe": "metrotestharness.exe"
 }
diff --git a/mozharness/mozilla/testing/talos.py b/mozharness/mozilla/testing/talos.py
index 71ee9f2..7b1cda1 100755
--- a/mozharness/mozilla/testing/talos.py
+++ b/mozharness/mozilla/testing/talos.py
@@ -92,6 +92,12 @@ class Talos(TestingMixin, MercurialScript):
            "dest": "branch",
            "help": "Graphserver branch to report to"
            }],
+        [["--metro-immersive"],
+          {"action": "store_true",
+           "dest": "metro_immersive",
+           "default": False,
+           "help": "Tells windows 8 machines to run tests with Metro Browser"
+           }],
         [["--system-bits"],
           {"action": "store",
            "dest": "system_bits",
@@ -152,6 +158,20 @@ class Talos(TestingMixin, MercurialScript):
         if 'run-tests' in self.actions:
             self.preflight_run_tests()
 
+    def query_abs_dirs(self):
+        c = self.config
+        if self.abs_dirs:
+            return self.abs_dirs
+        abs_dirs = super(Talos, self).query_abs_dirs()
+        dirs = {}
+        dirs['abs_test_dir'] = os.path.join(abs_dirs['abs_work_dir'],
+                                            'tests')
+        dirs['abs_metro_harness_dir'] = os.path.join(dirs['abs_test_dir'],
+                                                     c.get('metro_harness_dir', '')
+        abs_dirs.update(dirs)
+        self.abs_dirs = abs_dirs
+        return self.abs_dirs
+
     def query_talos_json_url(self):
         """Hacky, but I haven't figured out a better way to get the
         talos json url before we install the build.
@@ -481,6 +501,18 @@ class Talos(TestingMixin, MercurialScript):
             self.mkdir_p(os.path.dirname(manifest_target))
             self.copyfile(manifest_source, manifest_target)
 
+    def install(self):
+        """decorates TestingMixin.install() to handle win metro browser"""
+        c = self.config
+        dirs = self.query_abs_dirs()
+        super(Talos, self).install()
+        if c.get('metro_immersive'):
+            # overwrite self.binary_path set from TestingMixin.install()
+            self.binary_path = os.path.join(dirs['abs_metro_harness_dir'],
+                                            c.get('metro_test_harness_exe')
+            if not os.path.exists(self.binary_path):
+                self.fatal("metrotestharness executable could not be found")
+
     def preflight_run_tests(self):
         if not self.query_tests():
             self.fatal("No tests specified; please specify --tests")
